crates/contracts/solidity/Trader.sol

```diff
@@ -13,78 +13,72 @@ contract Trader {
-            receiver: address(0),
-            sellAmount: clearingPrices[tokens.length - 1],
-            buyAmount: clearingPrices[0],
-            validTo: type(uint32).max,
-            appData: bytes32(0),
-            feeAmount: 0,
-            flags: 0x40, // EIP-1271
-            executedAmount: 0,
-            signature: abi.encodePacked(address(this))
-        });
+    // The `Trader` contract gets deployed on the `from` address of the quote.
+    // Since the `from` address might be a safe or other smart contract we still
+    // need to make the `Trader` behave as the original `from` would have in
+    // case some custom interactions rely on that behvavior.
+    // To do that we simply implement fallback handlers that do delegate calls
+    // to the original implementation.
+    fallback() external payable {
+        _fallback();
+    }
+    receive() external payable {
+        _fallback();
+    }
+    function _fallback() private {
+        bytes memory rdata = TRADER_IMPL.doDelegatecall(msg.data);
+        assembly { return(add(rdata, 32), mload(rdata)) }
+    }
 
-        gasUsed = address(SETTLEMENT).doMeteredCallNoReturn(
-            abi.encodeCall(
-                SETTLEMENT.settle,
-                (tokens, clearingPrices, trades, interactions)
-            )
-        );
+    /// @dev Prepares everything needed by the trader for successfully executing the swap.
+    /// This includes giving the required approval, wrapping the required ETH (if needed)
+    /// and warming the needed storage for sending native ETH to smart contracts.
+    /// @param sellToken - token being sold by the trade
+    /// @param sellAmount - expected amount to be sold according to the quote
+    /// @param nativeToken - ERC20 version of the chain's native token
+    /// @param receiver - address that will receive the bought tokens
+    function prepareSwap(
+        address sellToken,
+        uint256 sellAmount,
+        address nativeToken,
+        address payable receiver
+    ) external {
+        require(!_alreadyCalled, "prepareSwap can only be called once");
+        _alreadyCalled = true;
 
-        for (uint256 i; i < tokens.length; ++i) {
-            traderBalances[i] += IERC20(tokens[i]).balanceOf(address(this)).toInt();
-            settlementBalances[i] += IERC20(tokens[i]).balanceOf(address(SETTLEMENT)).toInt();
+        if (sellToken == nativeToken) {
+            uint256 availableBalance = IERC20(sellToken).balanceOf(address(this));
+            if (availableBalance < sellAmount) {
+                // Simulate wrapping the missing `ETH` so the user doesn't have to spend gas
+                // on that just to get a quote. If they are happy with the quote and want to
+                // create an order they will actually have to do the wrapping, though.
+                (bool sent, ) = nativeToken.call{value: sellAmount - availableBalance}("");
+                require(sent, "'from' address does not have enough (W)ETH balance");
```

Comment:
I _think_ you should be able to do:
```suggestion
                nativeToken.deposit{value: sellAmount - availableBalance}();
```

This will automatically revert on failure.